api/handlers.go

```diff
@@ -497,42 +497,24 @@ func getContainerJSON(c *context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	client, scheme, err := container.Engine.HTTPClientAndScheme()
+	con, err := container.Engine.InspectContainer(container.ID)
 	if err != nil {
 		httpError(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
 
-	resp, err := client.Get(scheme + "://" + container.Engine.Addr + "/containers/" + container.ID + "/json")
-	container.Engine.CheckConnectionErr(err)
-	if err != nil {
-		httpError(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
+	con.Node = container.Engine.EngineToContainerNode()
 
-	// cleanup
-	defer resp.Body.Close()
-
-	data, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		httpError(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-
-	n, err := json.Marshal(container.Engine)
-	if err != nil {
-		httpError(w, err.Error(), http.StatusInternalServerError)
-		return
+	for _, portBindings := range con.NetworkSettings.Ports {
+		for key, portBinding := range portBindings {
+			if portBinding.HostIP == "0.0.0.0" {
```

Comment:
Swarm should run under IPv4 or IPv6. Here you are trying to check if address is unspecified. Golang has [IsUnspecified](https://golang.org/src/net/ip.go) supporting both IPv4 and IPv6.